[Nam Harin] iP

harin0826 opened the pull request:
## Update for Duke

Totally relevant quote:
>"Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

This is an update for Duke program.

Here are some of the added features:
- Adding tasks
- Marking tasks
- Removing tasks

Tasks are also classified into one of the following classes:
1. Todo - Simple version with just a task description
2. Deadline - Task with deadlines
3. Event - Task with start, end dates

`main` method is as following:
```java
public static void main(String[] args) {

        String home = System.getProperty("user.home");
        Duke duke = new Duke(home);
        System.out.println("Loading complete.\n");

        duke.ui.printHelloMsg();

        duke.run();

        duke.terminate();

} 
```
so it **also handles file systems of different** OS. :smile:


Things still left to implement:

- [x] Week 4's iP tasks
- [ ] Week 5's iP tasks

Commit:
Add Gradle support

Commit:
add String dividor
modified greeting

Commit:
Renamed String dividor to div_open
add String div_close

Commit:
Separated greetings message from logo

Commit:
Added echo and exit

Commit:
Added String array list

Commit:
Modified dividor names

Commit:
Add list command

Commit:
Added add, fixed bugs for wrong list numbering

Commit:
Added class Task

Commit:
Added toString method

Commit:
Changed if-else to switch cases

Commit:
Fixed Duke to utilize class Task

Commit:
Completed mark and unmark commands

Commit:
Added classes Todo, Deadline and Event

Commit:
Added commands todo, deadline, event
Added print for number of tasks

Commit:
Fixed minor formatting issue to exclude blank character

Commit:
Automated Text UI testing

Commit:
Added DukeException class

Commit:
Attempt to separate commands with methods

Commit:
Added parser method

Commit:
Set up necessary methods to be called after parser

Commit:
Draft for todo, deadline, event methods

Commit:
Fixed bugs where taskName is not displayed fully

Commit:
Added more exception descriptions

Commit:
Changed list to ArrayList

Commit:
Modified program to utilize ArrayList, added unknown command exception

Commit:
Added delete command and its test cases

Commit:
Modified Task class to use enums as task status, added new status "DOING"

Commit:
Adjust case indentation to follow java convention

Commit:
Add basic file support to open /data/data.txt

Need a system to save and load tasks for Duke.

Java NIO Paths class is used to load up directory for Duke's data
and load data.txt file to load any existing tasks from previous
execution of the program.

Structure for data.txt on how it will be saved is implemented later.

Commit:
Add saveFile method that saves current list of tasks to data.txt

Need a way to save current list of tasks to be loaded in future.

Update Task class and all of its subclasses to tokenise the current
tasks in the list. Write this to data.txt and save it.

Commit:
Add switch case in loadFile method to parse data read

Commit:
Add new constructor for Task class

data.txt is able to contain information about task's status. This
constructor allows the information to be loaded in as well when
Duke runs and load data in from data.txt

Commit:
Complete save and load function

Update constructors for all subclasses of Task class to allow
task status to be loaded from previous data. Duke program is also
updated to allow data saving every update of tasks.

Need to work out a way to make duke only update added/deleted
tasks as it overwrites the whole data.txt, costing running time.

Commit:
Add experimental LocalDate when parsing for Deadline class

ChinJunAn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,221 @@
+
+        switch(command[0]) {
+
+            case "list":
```
This violates the coding standard listed in class! The "case" statement must have the same indent at the "switch" statement.

ChinJunAn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,221 @@
+        }
+    }
+
+    public static void markTask(String arg) throws DukeException {
```
I think you can separate these functions into different classes to encourage OOP

ChinJunAn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,221 @@
+    private static final String DIV_CLOSE = "____________________________________________________________\n";
+
+    public static ArrayList<Task> list = new ArrayList<>();
+    public static int listNum = 1;
```
I think you can separate this into a TaskList class to allow OOP

ChinJunAn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,221 @@
+
+            case "unmark":
+                if (command.length != 2) {
+                    throw new DukeException("Please check the number of your arguments!");
```
I think you could improve the exception into more specific ones.

ChinJunAn reviewed (commented):
Areas for improvement!

Commit:
Add LocalDate variable to Deadline

Implementing LocalDate within Deadline class is simpler and more
OOP, hence it was moved off the parser within Duke class over to
the Deadline class.

Commit:
Implement LocalDate class to store dates within Task classes

Expanded on top of previous commit and implemented LocalDate within
Event class too. Added extra format method in toString method to
display dates in different format.

Commit:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/Duke.java

Commit:
Add new classes for more OOP implementation

Commit:
Implement Storage class to open and load data file

Commit:
Implement TaskList class to manage tasks

Commit:
Make the code more OOP

Methods are extracted into respective classes to support more OOP.

Commit:
Fix bugs where wrong date input format would crash the program

Not all constructors for tasks were updated and loading would result
in tasks not having proper LocalDate attribute.

Applied relevant changes to load tasks correctly with proper dates.

Commit:
Pack all classes under duke package

dunliang0513 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,9 @@
+package duke;
+
+public class DukeException extends Exception {
+
+    public DukeException(String message) {
+        super(message);
+    }
+
+}
```
I feel that you can specify different exceptions to handle different cases of errors, such as InvalidCommandExceptions, InvalidDeadlineExceptions, and InvalidDescriptionExceptions.

dunliang0513 commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,23 @@
+package duke;
+
+public class Todo extends Task {
+
+    public Todo(String description) {
+        super(description);
+    }
+
+    public Todo(String description, String status) {
+        super(description, status);
+    }
+
+    @Override
+    public String toString() {
+        return "[T]" + super.toString();
+    }
+
+    @Override
+    public String toData() {
+        return "T|" + super.toData();
+    }
+
+}
```
I think the Todo.java, Event.java, Task.java, and Deadline.java can be classified under a package name Tasks.

dunliang0513 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,34 @@
+package duke;
+
```
I think that you can apply some ENUM instances in the Ui class.
```suggestion
public static final String LINE = "------------------------------\n"
```

dunliang0513 reviewed (commented):
Most of the code standard is closely adhered to, but a few things need to take note: 
1. no Javadoc is provided, 
2. classify the classes into different packages instead of all under the duke package

Overall I think the logic and the flow are good manners. I think you have done a pretty good job, but need to bulk up on your progress.

Commit:
Merge branch 'master' into add-gradle-support

Commit:
Add JUnit tests for Todo and Storage

Commit:
Update build.gradle for shadowjar

Commit:
Change build.gradle's mainClassName

Commit:
Add JavaDoc comments for all public methods

Commit:
Tweak the code to comply with a coding standard

Commit:
Add find method that prints all tasks with matching keyword

Commit:
Tweak code to comply with coding standard checked by CheckStyle

Commit:
Add JavaFX basics for GUI implementation

Commit:
Add GUI to Duke

Greeting message and closing message still needs to be implemented.

Commit:
Add greeting message and handled exiting program

Greeting message was added in the initialize method of MainWindow.
Exiting program was handled by multithreading a ExitThread and
calling System.exit(0).

Commit:
Allow generation of working JAR file

Previously shadowjar for Gradle would exprot a JAR file that keeps
crashing. This was caused by the resource not being found because
of the file name. Resources under out extends the file with .PNG
instead of .png which caused the error.

New JAR file will properly load resources by looking for images as
.PNG extension.

Future management of resources must be carefully handled to consider
file extensions too.

Commit:
Add dummy change for branching

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Add dummy change for branching

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Add dummy change for branching to support grading system

Commit:
Merge branch 'branch-level-9'

Commit:
Add dummy change for branching to support grading system

Commit:
Merge branch 'branch-level-10'

Commit:
Add assertions for number of tasks in TaskList

Loading tasks from previous iterations create new TaskList object
that increments numOfTask variable as it loads task. This could be
affected if saved tasks are corrupted or task deletion happens when
despite having no tasks left. Negative numOfTasks could result in
errors when saving tasks.

This assertion ensures that numOfTasks are never negative and load
and save of tasks happen properly.

Commit:
Improve code quality

TaskList had many methods running for loops of supplied arguments
that essentially does the same thing, adding unnecessary
complications to the code.

Cleaned up and refactored with extract method while also improving
some of the comments within the code to improve readibility.

Commit:
Modify find method to utilize Stream

Current implementation loops through the list of tasks to check if
its description contains the argument, then add it into list of
found tasks, then prints out based on the found tasks.

This can be efficiently implemented with Streams.

Printing all tasks in lists can also be modified to use Streams
in the future.

harin0826 committed:
Merge pull request #2 from harin0826/branch-A-Assertions

Add assertions for number of tasks in TaskList

harin0826 committed:
Merge pull request #1 from harin0826/branch-A-CodeQuality

Improve code quality

harin0826 committed:
Merge pull request #3 from harin0826/branch-A-Streams

Modify find method to utilize Stream

harin0826 committed:
Create gradle.yml for Continuous Integration

Commit:
Add tag variable to Task and modify find function to support tags

Commit:
Update find function to improve support for tags

Commit:
Add more support for find function

Tag were not getting saved or properly assigned on creation and
loading of tasks. This was because String.split method would return
String array of size 1 if regex is not found.

Add check for array size and properly assign empty tag if tag does
not exist.

Commit:
Add JavaDoc for private find methods to add clarification

Commit:
Add UserGuide and Screenshot for product website

Commit:
Update UserGuide with tag command
